docs/developer-guide/ceps/proposed/cep-004-module-structure.rst

```diff
@@ -0,0 +1,79 @@
+.. _cep-003:
+
+************************************************************
+CEP 4 - Changing the module structure of the ctapipe package
+************************************************************
+
+* Status: draft
+* Discussion: NA
+* Date accepted: NA
+* Last revised: 2024-01-08
+* Author: Maximilian Linhoff
+* Created: 2024-01-08
+
+
+Abstract
+========
+
+``ctapipe``'s module structure has organically grown over its development
+in the last couple of years and has some suboptimal properties.
+
+* Large inter-dependencies between modules, sometimes even circular. E.g. importing ``ctapipe.calib`` will import ``ctapipe.image`` because the image extraction code lives in ``ctapipe.image``.
+* Deep hierarchies, resulting in issues with references in the sphinx documentation due to the same class being documentated at more than two locations.
+
+A better structure would resolve these issues and would also make it easier to
+assign maintainers / code owners to specific submodules.
+
+
+Proposed new structure
+======================
+
+In the new structure, we want to use a flatter hierarchy of smaller modules,
+that will resolve the above mentioned issues.
+
+We expect no disadvantages other than:
+* Breaking compatibility of imports once
+* More namespaces for developers to remember
+
+The second point should be offset by the more logical structure of the namespaces.
+
+.. code::
+
+   ctapipe
+        .atmosphere
+        .camera_calibration
+        .config
+        .coordinates
+        .core
+        .data_model
+        .data_volume_reduction
+        .gain_selection
+        .image_cleaning
+        .image_features
+        .instrument
+        .muon
+        .processors
+        .reconstruction
+        .tools
+        .visualization
+
+
+Looking at it from the current structure, we'd have the following changes:
+
+- ``.image`` is split into ``.image_cleaning`` and ``.image_features``
+- ``.calib`` is split into ``.camera_calibration``, ``.data_volume_reduction``, ``.gain_selection``
+- ``.image.extractor`` is moved to ``.camera_calibration``
+- ``.image.muon`` is moved to ``.muon``
+- ``.containers`` is renamed to ``.data_model``
```

Comment:
Do you mean `core.container.py` is moved to its own module?

Reply:
No, containers.py as written

Reply:
Ah, I see! 

Then I would add I think it is quite confusing that some of the modules are in their own sub-dirs and others are just files hanging out in the root `src/` directory. 

I suggest we put everything that is supposed to be a proper module into its own subdirectory, leaving only things like `version.py` and `conftest.py` in the root. This also makes a natural place for a dedicated `tests/` subdirectory for things like atmosphere and containers instead of having a central one collecting test for various, but not all, modules.